pyro/distributions/torch_wrapper.py

```diff
@@ -13,36 +13,41 @@ class TorchDistribution(Distribution):
     `torch.distributions.Distribution <http://pytorch.org/docs/master/_modules/torch/distributions.html#Distribution>`_
     """
 
-    def __init__(self, torch_dist, x_shape, event_dim, log_pdf_mask=None):
+    def __init__(self, torch_dist, log_pdf_mask=None, extra_event_dims=0):
         super(TorchDistribution, self).__init__()
         self.torch_dist = torch_dist
         self.log_pdf_mask = log_pdf_mask
-        self._x_shape = x_shape
-        self._event_dim = event_dim
+        self.extra_event_dims = extra_event_dims
 
-    def batch_shape(self, x=None):
-        x_shape = [] if x is None else x.size()
-        shape = torch.Size(broadcast_shape(x_shape, self._x_shape, strict=True))
-        event_start = len(shape) - self._event_dim
-        return shape[:event_start]
+    def batch_shape(self):
+        if self.extra_event_dims == 0:
+            return self.torch_dist.batch_shape
+        batch_dims = len(self.torch_dist.batch_shape) - self.extra_event_dims
+        return self.torch_dist.batch_shape[:batch_dims]
 
     def event_shape(self):
-        event_start = len(self._x_shape) - self._event_dim
-        return self._x_shape[event_start:]
+        if self.extra_event_dims == 0:
+            return self.torch_dist.event_shape
+        batch_dims = len(self.torch_dist.batch_shape) - self.extra_event_dims
+        shape = self.torch_dist.batch_shape + self.torch_dist.event_shape
+        return shape[batch_dims:]
 
     def sample(self, sample_shape=torch.Size()):
         if self.reparameterized:
             return self.torch_dist.rsample(sample_shape)
         else:
             return self.torch_dist.sample(sample_shape)
 
-    def batch_log_pdf(self, x):
-        batch_log_pdf_shape = self.batch_shape(x) + (1,)
-        log_pxs = self.torch_dist.log_prob(x)
-        batch_log_pdf = torch.sum(log_pxs, -1).contiguous().view(batch_log_pdf_shape)
+    def log_prob(self, x):
+        log_prob = self.torch_dist.log_prob(x)
+        for _ in range(self.extra_event_dims):
+            log_prob = log_prob.sum(-1)
         if self.log_pdf_mask is not None:
-            batch_log_pdf = batch_log_pdf * self.log_pdf_mask
-        return batch_log_pdf
+            # Prevent accidental broadcasting of log_prob tensor,
+            # e.g. (64, 1), (64,) --> (64, 64)
+            assert len(self.log_pdf_mask.size()) <= len(log_prob.size())
```

Comment:
:+1: Good idea!

Reply:
This was an issue with the AIR tutorial, so thought of adding this as a check.